pkg/health/exec.go

```diff
@@ -43,15 +44,30 @@ func (e *ExecHealthChecker) HealthCheck(podFullName string, podUID types.UID, st
 	if container.LivenessProbe.Exec == nil {
 		return Unknown, fmt.Errorf("missing exec parameters")
 	}
-	data, err := e.runner.RunInContainer(podFullName, podUID, container.Name, container.LivenessProbe.Exec.Command)
-	glog.V(1).Infof("container %s health check response: %s", podFullName, string(data))
-	if err != nil {
-		return Unknown, err
+
+	type containerStatusResult struct {
+		st  Status
+		err error
 	}
-	if strings.ToLower(string(data)) != defaultHealthyOutput {
+
+	ch := make(chan containerStatusResult, 1)
+	go func() {
+		data, err := e.runner.RunInContainer(podFullName, podUID, container.Name, container.LivenessProbe.Exec.Command)
+		glog.V(1).Infof("container %s health check response: %s", podFullName, string(data))
+		if err != nil {
+			ch <- containerStatusResult{Unknown, err}
+		}
+		if strings.ToLower(string(data)) != defaultHealthyOutput {
+			ch <- containerStatusResult{Unhealthy, nil}
+		}
+		ch <- containerStatusResult{Healthy, nil}
+	}()
+	select {
+	case statusResult := (<-ch):
+		return statusResult.st, statusResult.err
+	case <-time.After(timeout(container)):
```

Comment:
I predict a leak - a health-check command that locks up and times out, followed by a container restart, and repeat.  Is there no way to do a subprocess that we can reliably kill?  Can we fork() in go, and set an alarm signal?